collector/exporter/otelarrowexporter/internal/arrow/bestofn.go

```diff
@@ -0,0 +1,144 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package arrow // import "github.com/open-telemetry/otel-arrow/collector/exporter/otelarrowexporter/internal/arrow"
+
+import (
+	"context"
+	"runtime"
+	"sort"
+)
+
+// bestOfNPrioritizer is a prioritizer that selects a less-loaded stream to write.
+// https://smallrye.io/smallrye-stork/1.1.1/load-balancer/power-of-two-choices/
+type bestOfNPrioritizer struct {
+	doneCancel
+
+	// input from the pipeline, as processed data with headers and
+	// a return channel for the result.  This channel is never
+	// closed and is buffered.  At shutdown, items of telemetry can
+	// be left in this channel, but users are expected to complete
+	// their requests before calling shutdown (and the collector's
+	// graph package ensures this).
+	input chan writeItem
+
+	// state tracks the work being handled by all streams.
+	state map[*streamWorkState]struct{}
+
+	// N is the number of streams to consder in each decision.
+	N int
+
+	// loadFunc is the load function.
+	loadFunc loadFunc
+}
+
+type loadFunc func(*streamWorkState) float64
+
+type streamSorter struct {
+	work *streamWorkState
+	load float64
+}
+
+var _ streamPrioritizer = &bestOfNPrioritizer{}
+
+func newBestOfNPrioritizer(dc doneCancel, N, numStreams int, lf loadFunc) (*bestOfNPrioritizer, []*streamWorkState) {
+	var sws []*streamWorkState
+	state := map[*streamWorkState]struct{}{}
+
+	for i := 0; i < numStreams; i++ {
+		ws := &streamWorkState{
+			waiters: map[int64]chan<- error{},
+			toWrite: make(chan writeItem, 1),
+		}
+
+		sws = append(sws, ws)
+		state[ws] = struct{}{}
+	}
+
+	lp := &bestOfNPrioritizer{
+		doneCancel: dc,
+		input:      make(chan writeItem, runtime.NumCPU()),
+		state:      state,
+		N:          N,
+		loadFunc:   lf,
+	}
+
+	for i := 0; i < len(lp.state); i++ {
+		// TODO It's not clear if/when the the prioritizer can
+		// become a bottleneck.
+		go lp.run()
+	}
+
+	return lp, sws
+}
+
+func (lp *bestOfNPrioritizer) downgrade(ctx context.Context) {
+	for ws := range lp.state {
+		go drain(ws.toWrite, ctx.Done())
+	}
+}
+
+func (lp *bestOfNPrioritizer) sendOne(item writeItem, tmp []streamSorter) {
+	stream := lp.streamFor(item, tmp)
+	writeCh := stream.toWrite
+	select {
+	case writeCh <- item:
+		return
```

Comment:
Do you need to send `item.errCh <- nil` here to signal to waiter before returning?

Reply:
In this case, the item of work is moving into the stream for processing, so this is the only branch where item.errCh is not used. Otherwise, all paths lead to `errCh <- ErrStreamRestarting` below.